---
title: "Actinopterygii whole genomes analyses"
author: "Alina G. Mikhailova"
date: "2022-10-25"
output:    
   html_document:
      toc: true
      toc_depth: 3
      toc_float: true
      number_section: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo=FALSE, message=FALSE}
rm(list=ls(all=TRUE))
if (!require(caper)) install.packages("caper")
if (!require(geiger)) install.packages("geiger")
if (!require(ggpubr)) install.packages("ggpubr")
if (!require(dplyr)) install.packages("dplyr")
library(caper)
library(geiger)
library(ggpubr)
library(dplyr)
```

# Reading whole genomes database for Actinopterigii (LIGHT CHAIN NOTATION)

```{r}
AllSynNuc = read.table("../data/1raw/AllGenesCodonUsageNoOverlap.txt", header = TRUE, sep = '\t')
AllSynNuc = AllSynNuc[AllSynNuc$Gene != 'ND6',]
```

# Obtaining neutral nucleotide fractions in whole genomes
```{r message=FALSE}
AllSynNuc = aggregate(list(AllSynNuc$NeutralA,AllSynNuc$NeutralT,AllSynNuc$NeutralG,AllSynNuc$NeutralC), by = list(AllSynNuc$Species), FUN = sum)
names(AllSynNuc) = c('Species','NeutralA','NeutralT','NeutralG','NeutralC')
AllSynNuc$FrT.heavy = AllSynNuc$NeutralA / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC)
AllSynNuc$FrA.heavy = AllSynNuc$NeutralT / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
AllSynNuc$FrC.heavy = AllSynNuc$NeutralG / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
AllSynNuc$FrG.heavy = AllSynNuc$NeutralC / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
```

# Merging neutral freqs with time of maturation and temperature
```{r message=FALSE}
tm = read.table('../Data/1raw/FishBaseTemperature.txt', header=T)
tm = tm %>% group_by(Species) %>% summarise(temperature = median(Temperature))
matur = read.table('../Data/1raw/FishBaseMaturity_Tm.txt', header=T)
matur = matur %>% group_by(Species) %>% summarise(matur_tm = median(Tm))
AllSynNuc = merge(AllSynNuc, tm, by='Species', all.x=T)
AllSynNuc = merge(AllSynNuc, matur, by='Species', all.x=T)
```

# Correlations of neutral freqs with temperature
## Correlation of neutral freqs with temperature
```{r message=FALSE, warning=FALSE}  
nrow(AllSynNuc[!is.na(AllSynNuc$FrA.heavy) & !is.na(AllSynNuc$temperature),])
cor.test(AllSynNuc$temperature,AllSynNuc$FrA.heavy, method = "spearman")
cor.test(AllSynNuc$temperature,AllSynNuc$FrT.heavy, method = "spearman")
cor.test(AllSynNuc$temperature,AllSynNuc$FrG.heavy, method = "spearman")
cor.test(AllSynNuc$temperature,AllSynNuc$FrC.heavy, method = "spearman")  
```

## Correlation of neutral freqs with time of maturation
```{r message=FALSE, warning=FALSE}  
nrow(AllSynNuc[!is.na(AllSynNuc$FrA.heavy) & !is.na(AllSynNuc$matur_tm),])
cor.test(AllSynNuc$matur_tm,AllSynNuc$FrA.heavy, method = "spearman")
cor.test(AllSynNuc$matur_tm,AllSynNuc$FrT.heavy, method = "spearman")
cor.test(AllSynNuc$matur_tm,AllSynNuc$FrG.heavy, method = "spearman")
cor.test(AllSynNuc$matur_tm,AllSynNuc$FrC.heavy, method = "spearman")  
```

## GAskew, TCskew, STG-SACSkew
```{r message=FALSE, warning=FALSE}
AllSynNuc$TtoCSkew.heavy = (AllSynNuc$FrT.heavy-AllSynNuc$FrC.heavy)/(AllSynNuc$FrT.heavy+AllSynNuc$FrC.heavy)
AllSynNuc$GtoASkew.heavy = (AllSynNuc$FrG.heavy-AllSynNuc$FrA.heavy)/(AllSynNuc$FrG.heavy+AllSynNuc$FrA.heavy)

nrow(AllSynNuc[!is.na(AllSynNuc$GtoASkew.heavy) & !is.na(AllSynNuc$temperature),])
cor.test(AllSynNuc$temperature,AllSynNuc$TtoCSkew.heavy, method = "spearman")
cor.test(AllSynNuc$temperature,AllSynNuc$GtoASkew.heavy, method = "spearman")
nrow(AllSynNuc[!is.na(AllSynNuc$GtoASkew.heavy) & !is.na(AllSynNuc$matur_tm),])
cor.test(AllSynNuc$matur_tm,AllSynNuc$TtoCSkew.heavy, method = "spearman")
cor.test(AllSynNuc$matur_tm,AllSynNuc$GtoASkew.heavy, method = "spearman")

AllSynNuc$TG = AllSynNuc$FrT.heavy+AllSynNuc$FrG.heavy
AllSynNuc$AC = AllSynNuc$FrA.heavy+AllSynNuc$FrC.heavy
AllSynNuc$TG_ACSkew.heavy = (AllSynNuc$TG-AllSynNuc$AC)/(AllSynNuc$TG+AllSynNuc$AC)

```

# Figures
```{r message=FALSE, warning= FALSE}
AllSynNuc = AllSynNuc[!is.na(AllSynNuc$temperature),]; N = as.character(paste("N", nrow(AllSynNuc), sep="")) #delete NA
pdf("../Figures/WholeGenomeData.Actinopterygii.pdf")
f1= ggscatter(AllSynNuc, x = "temperature", y = "FrA.heavy",
          color = "#e61a0b", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of AH")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
f1
f2=ggscatter(AllSynNuc, x = "temperature", y = "FrC.heavy",
          color = "#009414", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of CH")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
f2
f3=ggscatter(AllSynNuc, x = "temperature", y = "FrG.heavy",
          color = "#5c5c5c", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of GH")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
f3
f4=ggscatter(AllSynNuc, x = "temperature", y = "FrT.heavy",
          color = "#0918e6", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of TH")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
f4
dev.off()
f1
f2
f3
f4
```

```{r message=FALSE, warning= FALSE}
pdf("../Figures/WholeGenomeData.Actinopterygii.GAskew.pdf", width = 7, height = 7)
f5=ggscatter(AllSynNuc, x = "temperature", y = "GtoASkew.heavy",
          color = "#c99bc9", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Mean annual water temperature, ?C",  ylim=c(0.1, 0.65), xlim=c(0,30))+stat_cor(method = "spearman", label.x = 2.5, label.y = 0.64, aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
f5
dev.off()
f5
```

# LM
```{r}
summary(lm(GtoASkew.heavy ~ temperature, data = AllSynNuc))
summary(lm(TtoCSkew.heavy ~ temperature, data = AllSynNuc))
summary(lm(TG_ACSkew.heavy ~ temperature, data = AllSynNuc))
summary(lm(GtoASkew.heavy ~ matur_tm, data = AllSynNuc))
summary(lm(TtoCSkew.heavy ~ matur_tm, data = AllSynNuc))
summary(lm(TG_ACSkew.heavy ~ matur_tm, data = AllSynNuc))


nrow(AllSynNuc[!is.na(AllSynNuc$TG_ACSkew.heavy) & !is.na(AllSynNuc$temperature)& ! is.na(AllSynNuc$matur_tm),])

summary(lm(FrA.heavy ~ scale(temperature)+scale(matur_tm), data = AllSynNuc))
summary(lm(FrA.heavy ~ log2(temperature + 2)*log2(matur_tm), data = AllSynNuc))  
summary(lm(FrA.heavy ~ log2(temperature + 2)+log2(matur_tm), data = AllSynNuc))
summary(lm(FrT.heavy ~ log2(temperature + 2)+log2(matur_tm), data = AllSynNuc)) # strong
summary(lm(FrA.heavy ~ log2(temperature + 2)+log2(matur_tm), data = AllSynNuc)) # strong



summary(lm(TG_ACSkew.heavy ~ log2(temperature + 2)+log2(matur_tm), data = AllSynNuc))
summary(lm(TG_ACSkew.heavy ~ scale(temperature + 2)+scale(matur_tm), data = AllSynNuc))# ###PICS


summary(lm(GtoASkew.heavy ~ scale(temperature + 2)+scale(matur_tm), data = AllSynNuc))
summary(lm(TtoCSkew.heavy ~ scale(temperature + 2)+scale(matur_tm), data = AllSynNuc))
summary(lm(GtoASkew.heavy ~ scale(temperature + 2)*scale(matur_tm), data = AllSynNuc))
summary(lm(TtoCSkew.heavy ~ scale(temperature + 2)*scale(matur_tm), data = AllSynNuc))
nrow(AllSynNuc[!is.na(AllSynNuc$temperature) & !is.na(AllSynNuc$matur_tm),])


summary(lm(temperature ~ GtoASkew.heavy +TtoCSkew.heavy, data = AllSynNuc))
summary(lm(matur_tm ~ GtoASkew.heavy +TtoCSkew.heavy, data = AllSynNuc))
summary(lm(temperature ~ scale(GtoASkew.heavy) +scale(TtoCSkew.heavy), data = AllSynNuc))
summary(lm(matur_tm ~ scale(GtoASkew.heavy) +scale(TtoCSkew.heavy), data = AllSynNuc))

``` 

# PGLS
```{r warning=FALSE}
tree = read.tree('../Data/1raw/mtalign.aln.treefile.rooted')

row.names(AllSynNuc) = AllSynNuc$Species
tree_pruned = treedata(tree, AllSynNuc, sort=T, warnings=T)$phy

data<-as.data.frame(treedata(tree_pruned, AllSynNuc, sort=T, warnings=T)$data)
data$Species = as.character(data$Species)
data$TG_ACSkew.heavy = as.numeric(as.character(data$TG_ACSkew.heavy))
data$GtoASkew.heavy = as.numeric(as.character(data$GtoASkew.heavy))
data$TtoCSkew.heavy = as.numeric(as.character(data$TtoCSkew.heavy))
data$temperature = as.numeric(as.character(data$temperature))
data$matur_tm = as.numeric(as.character(data$matur_tm))
data_comp <- comparative.data(tree_pruned, data, Species, vcv=TRUE)


summary(pgls(GtoASkew.heavy ~ temperature, data_comp, lambda="ML"))
summary(pgls(TtoCSkew.heavy ~ temperature, data_comp, lambda="ML"))
summary(pgls(TG_ACSkew.heavy ~ temperature, data_comp, lambda="ML"))

summary(pgls(GtoASkew.heavy ~ log2(temperature + 2) + log2(matur_tm), data_comp, lambda="ML"))
summary(pgls(TtoCSkew.heavy ~ log2(temperature + 2) + log2(matur_tm), data_comp, lambda="ML"))
summary(pgls(TG_ACSkew.heavy ~ log2(temperature + 2) + log2(matur_tm), data_comp, lambda="ML"))

summary(pgls(temperature ~ scale(GtoASkew.heavy) + scale(TtoCSkew.heavy), data_comp, lambda="ML"))
summary(pgls(matur_tm ~ scale(GtoASkew.heavy) + scale(TtoCSkew.heavy), data_comp, lambda="ML"))
```

#Figures
```{r, echo=FALSE, fig.show='hide'}
AllSynNuc = AllSynNuc[!is.na(AllSynNuc$matur_tm),]; N = as.character(paste("N", nrow(AllSynNuc), sep=""))
medianTm = median(AllSynNuc[!is.na(AllSynNuc$matur_tm),]$matur_tm)
AllSynNuc$Longevity = "Na"
AllSynNuc[AllSynNuc$matur_tm < medianTm,]$Longevity = "ShortLived"
AllSynNuc[AllSynNuc$matur_tm >= medianTm,]$Longevity = "LongLived"

plot(AllSynNuc[AllSynNuc$Longevity == "ShortLived",]$temperature, AllSynNuc[AllSynNuc$Longevity == "ShortLived",]$TG_ACSkew.heavy, col="#4da36c", xlab="Mean annual water temperature, C?", ylab="STG-SAC skew", ylim=c(0.1, 0.65), xlim=c(0,30), pch = 16)
abline((0.331911-0.049196), 0.006172, col="#4da36c", lwd = 2)
par(new=TRUE)
plot(AllSynNuc[AllSynNuc$Longevity == "LongLived",]$temperature, AllSynNuc[AllSynNuc$Longevity == "LongLived",]$TG_ACSkew.heavy, col="#42cbf5", xlab="", ylab="", ylim=c(0.1, 0.65), xlim=c(0,30), pch = 16)
abline(0.331911, 0.006172, col="#42cbf5", lwd = 2)
legend("bottomright", legend=c( "Short time of maturation","Long time of maturation"), col=c("#4da36c","#42cbf5"), pch = 16)
```