---
title: "Actinopterygii whole genomes analyses"
author: "Alina G. Mikhailova"
date: "2022-10-25"
output:    
   html_document:
      toc: true
      toc_depth: 3
      toc_float: true
      number_section: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo=FALSE, message=FALSE}
rm(list=ls(all=TRUE))
if (!require(caper)) install.packages("caper")
if (!require(geiger)) install.packages("geiger")
if (!require(ggpubr)) install.packages("ggpubr")
if (!require(dplyr)) install.packages("dplyr")
library(caper)
library(geiger)
library(ggpubr)
library(dplyr)
```

# Reading whole genomes database for Actinopterigii 

```{r}
AllSynNuc = read.table("../data/1raw/AllGenesCodonUsageNoOverlap.txt", header = TRUE, sep = '\t')
AllSynNuc = AllSynNuc[AllSynNuc$Gene != 'ND6',]
```

# Obtaining neutral nucleotide fractions in whole genomes
```{r message=FALSE}
AllSynNuc = aggregate(list(AllSynNuc$NeutralA,AllSynNuc$NeutralT,AllSynNuc$NeutralG,AllSynNuc$NeutralC), by = list(AllSynNuc$Species), FUN = sum)
names(AllSynNuc) = c('Species','NeutralA','NeutralT','NeutralG','NeutralC')
AllSynNuc$FrA = AllSynNuc$NeutralA / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC)
AllSynNuc$FrT = AllSynNuc$NeutralT / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
AllSynNuc$FrG = AllSynNuc$NeutralG / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
AllSynNuc$FrC = AllSynNuc$NeutralC / (AllSynNuc$NeutralA + AllSynNuc$NeutralT + AllSynNuc$NeutralG + AllSynNuc$NeutralC) 
```

# Merging neutral freqs with time of maturation and temperature
```{r message=FALSE}
tm = read.table('../Data/1raw/FishBaseTemperature.txt', header=T)
tm = tm %>% group_by(Species) %>% summarise(temperature = median(Temperature))
matur = read.table('../Data/1raw/FishBaseMaturity_Tm.txt', header=T)
matur = matur %>% group_by(Species) %>% summarise(matur_tm = median(Tm))
AllSynNuc = merge(AllSynNuc, tm, by='Species', all.x=T)
AllSynNuc = merge(AllSynNuc, matur, by='Species', all.x=T)
nrow(AllSynNuc[!is.na(AllSynNuc$temperature) & !is.na(AllSynNuc$FrT),])
```

# Correlations of neutral freqs with temperature
## Correlation of neutral freqs with temperature
```{r message=FALSE, warning=FALSE}  
cor.test(log2(AllSynNuc$temperature),AllSynNuc$FrA, method = "spearman")
cor.test(log2(AllSynNuc$temperature),AllSynNuc$FrT, method = "spearman")
cor.test(log2(AllSynNuc$temperature),AllSynNuc$FrG, method = "spearman")
cor.test(log2(AllSynNuc$temperature),AllSynNuc$FrC, method = "spearman")  
```

## Correlation of neutral freqs with time of maturation
```{r message=FALSE, warning=FALSE}  
cor.test(log2(AllSynNuc$matur_tm),AllSynNuc$FrA, method = "spearman")
cor.test(log2(AllSynNuc$matur_tm),AllSynNuc$FrT, method = "spearman")
cor.test(log2(AllSynNuc$matur_tm),AllSynNuc$FrG, method = "spearman")
cor.test(log2(AllSynNuc$matur_tm),AllSynNuc$FrC, method = "spearman")  
```

## Calculation of STG-SACSkew
```{r message=FALSE, warning=FALSE}
AllSynNuc$TG = AllSynNuc$FrT+AllSynNuc$FrG
AllSynNuc$AC = AllSynNuc$FrA+AllSynNuc$FrC
AllSynNuc$TG_ACSkew = (AllSynNuc$TG-AllSynNuc$AC)/(AllSynNuc$TG+AllSynNuc$AC);
AllSynNuc$AC_TGSkew = -(AllSynNuc$TG-AllSynNuc$AC)/(AllSynNuc$TG+AllSynNuc$AC); 
nrow(AllSynNuc[!is.na(AllSynNuc$AC_TGSkew) & !is.na(AllSynNuc$temperature),])
nrow(AllSynNuc[!is.na(AllSynNuc$AC_TGSkew) & !is.na(AllSynNuc$temperature) & !is.na(AllSynNuc$matur_tm),])
```

# Figures
```{r message=FALSE, warning= FALSE}
AllSynNuc = AllSynNuc[!is.na(AllSynNuc$temperature),]; N = as.character(paste("N", nrow(AllSynNuc), sep="")) #delete NA
ggscatter(AllSynNuc, x = "temperature", y = "FrT",
          color = "#e61a0b", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of A")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))

ggscatter(AllSynNuc, x = "temperature", y = "FrG",
          color = "#009414", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of C")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))

ggscatter(AllSynNuc, x = "temperature", y = "FrC",
          color = "#5c5c5c", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of G")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))

ggscatter(AllSynNuc, x = "temperature", y = "FrA",
          color = "#0918e6", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Median annual water temperature, ?C", ylab="Whole genome neutral fraction of T")+stat_cor(method = "spearman", aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
```

```{r message=FALSE, warning= FALSE}
ggscatter(AllSynNuc, x = "temperature", y = "AC_TGSkew",
          color = "#c99bc9", # Points color, shape and size
          add = "reg.line",  # Add regressin line
          add.params = list(color = "black", fill = "lightgray"), # Customize reg. line
          conf.int = TRUE, # Add confidence interval
          xlab="Mean annual water temperature, ?C", ylab="STG-SAC skew", ylim=c(0.1, 0.65), xlim=c(0,30))+stat_cor(method = "spearman", label.x = 2.5, label.y = 0.64, aes(label = paste(..r.label.., ..p.label.., ..N.., sep = "~`,`~")))
```

# LM
```{r}
summary(lm(AC_TGSkew ~ temperature, data = AllSynNuc))
summary(lm(AC_TGSkew ~ log2(temperature+2)+log2(matur_tm), data = AllSynNuc))
summary(lm(AC_TGSkew ~ scale(temperature+2)+scale(matur_tm), data = AllSynNuc))
``` 

# PGLS
```{r warning=FALSE}
#tree = read.tree('../Data/1raw/mtalign.aln.treefile.rooted')
tree = read.tree('../Data/1raw/Fishes_WholeGenomes_sp.nwk')

row.names(AllSynNuc) = AllSynNuc$Species
tree_pruned = treedata(tree, AllSynNuc, sort=T, warnings=T)$phy

data<-as.data.frame(treedata(tree_pruned, AllSynNuc, sort=T, warnings=T)$data)
data$Species = as.character(data$Species)
data$AC_TGSkew = as.numeric(as.character(data$AC_TGSkew))
data$temperature = as.numeric(as.character(data$temperature))
data$matur_tm = as.numeric(as.character(data$matur_tm))
data_comp <- comparative.data(tree_pruned, data, Species, vcv=TRUE)

summary(pgls(AC_TGSkew ~ temperature, data_comp, lambda="ML"))
summary(pgls(AC_TGSkew ~ scale(temperature), data_comp, lambda="ML"))
summary(pgls(log2(AC_TGSkew) ~ log2(temperature+2)+log2(matur_tm), data_comp, lambda="ML"))
```

#Figures
```{r}
AllSynNuc = AllSynNuc[!is.na(AllSynNuc$matur_tm),]; N = as.character(paste("N", nrow(AllSynNuc), sep=""))
medianTm = median(AllSynNuc[!is.na(AllSynNuc$matur_tm),]$matur_tm)
AllSynNuc$Longevity = "Na"
AllSynNuc[AllSynNuc$matur_tm < medianTm,]$Longevity = "ShortLived"
AllSynNuc[AllSynNuc$matur_tm >= medianTm,]$Longevity = "LongLived"

plot(AllSynNuc[AllSynNuc$Longevity == "ShortLived",]$temperature, AllSynNuc[AllSynNuc$Longevity == "ShortLived",]$AC_TGSkew, col="#4da36c", xlab="Mean annual water temperature, C?", ylab="STG-SAC skew", ylim=c(0.1, 0.65), xlim=c(0,30), pch = 16)
abline((0.331911-0.049196), 0.006172, col="#4da36c", lwd = 2)
par(new=TRUE)
plot(AllSynNuc[AllSynNuc$Longevity == "LongLived",]$temperature, AllSynNuc[AllSynNuc$Longevity == "LongLived",]$AC_TGSkew, col="#42cbf5", xlab="", ylab="", ylim=c(0.1, 0.65), xlim=c(0,30), pch = 16)
abline(0.331911, 0.006172, col="#42cbf5", lwd = 2)
legend("bottomright", legend=c( "Short time of maturation","Long time of maturation"), col=c("#4da36c","#42cbf5"), pch = 16)
```